database/misc/mongodb_replset.py

```diff
@@ -0,0 +1,218 @@
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            'login_user':     {'aliases': ['user']},
+            'login_password': {'aliases': ['password']},
+            'login_host':     {'aliases': ['host'], 'default': 'localhost'},
+            'login_port':     {'aliases': ['port'], 'default': 27017},
+            'hosts':          {'aliases': ['members'], 'required': True},
+            'replica_set':    {'aliases': ['replset'], 'required': True}
+        }
+    )
+
+    if not pymongo:
+        module.fail_json(msg='Python module "pymongo" must be installed.')
+
+    user, password, host, port, hosts, replset = (
+        module.params[k] for k in ['login_user', 'login_password', 'login_host',
+                                   'login_port', 'hosts', 'replica_set'])
+    nodes = split_hosts(hosts)
+
+    initiated = False
+    try:
+        try:
+            client = MongoReplicaSetClient(hosts, replicaSet=replset,
+                                           read_preference=ReadPreference.PRIMARY)
+            initiated = True
+        except ConfigurationError, e:
+            if 'is not a member of replica set' in e.message:
+                client = MongoClient(host, int(port), read_preference=ReadPreference.SECONDARY)
+            else:
+                module.fail_json(msg="Unable to connect: %s" % str(e))
+
+        if not user and not password:
+            credentials = load_mongocnf_creds()
+            if credentials is not False:
+                user = credentials['user']
+                password = credentials['password']
+        elif not password and user:
+            module.fail_json(msg='When supplying login arguments, both login_user and login_password must be provided')
+
+        if user and password:
+            try:
+                client.admin.authenticate(user, password)
+            except OperationFailure, e:
+                pass  # try to continue, maybe admin account is not set yet
+
+    except ConnectionFailure, e:
+        module.fail_json(msg='unable to connect to database: %s' % str(e))
+
+    changed = True
+    if initiated:
+        removed_hosts = client.hosts - set(nodes)
+        added_hosts = set(nodes) - client.hosts
+        if removed_hosts:
+            module.fail_json(msg="This module doesn't support members removing",
+                             members="'%s'" % removed_hosts)
+        elif added_hosts:
+            try:
+                add_members(client, nodes)
+            except OperationFailure, e:
+                module.fail_json(msg='Unable to add new members: %s' % str(e),
+                                 members="'%s'" % added_hosts)
+        else:
+            changed = False
+    else:
+        try:
+            replset_initiate(client, replset, nodes)
+        except OperationFailure, e:
+            module.fail_json(msg='Unable to initiate replica set: %s' % str(e))
+
+    module.exit_json(changed=changed)
```

Comment:
This `exit_json` contains very little info.  Some of the `fail_json` statements contain more info.